File: Lesson4-counting_elements/MissingInteger/mySol.c
// you can write to stdout for debugging purposes, e.g.
// printf("this is a debug message\n");

int solution(int A[], int N) {
    // write your code in C99 (gcc 6.2.0)
    int ret = 1;
    int i;

    if (N==1) return A[0]==1? 2: 1;

    int rangemap[100000-1] = {0};
    for (i=0; i<N; i++){
        if (A[i] < ret)
            continue;
        rangemap[A[i]-1] |= 1;
        if (A[i] == ret)
            ret = A[i];
    }
    for (i=(ret-1); i<(100000-1); i++){
        //printf("ret=%d, ",i+1);
        if (!rangemap[i])
            break;
        ret = i+1+1;
    }
    return ret;
}

File: Lesson5-prefix_sums/PassingCars/mySol.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def prefix_sums(A):
    P = [0]
    for i in range(len(A)):
        P += [P[i] + A[i]]
    return P

def solution(A):
    # write your code in Python 3.6
    ret = 0
    PreSum = prefix_sums(A)
    TotalSum = PreSum[len(A)]
    for i in range(len(A)):
        if 0 == A[i]:
            ret += TotalSum - PreSum[i]
            if ret > 1000000000:
                return -1
    return ret